torch/nn/parallel/distributed_c10d.py

```diff
@@ -0,0 +1,366 @@
+        if len(self.device_ids) > 1:
+            # intra-node parameter sync
+            result = broadcast_coalesced(self.modules_params_data[0],
+                                         self.device_ids,
+                                         self.broadcast_bucket_size)
+            for tensors, module_params_data in zip(result[1:], self.modules_params_data[1:]):
+                for tensor, param_data in zip(tensors, module_params_data):
+                    param_data.set_(tensor)
+
+        # module buffer sync
+        if self.broadcast_buffers:
+            if len(self.modules_buffers_data[0]) > 0:
+                # cross-node buffer sync
+                self._dist_broadcast_coalesced(self.modules_buffers_data[0],
+                                               self.broadcast_bucket_size)
+                if len(self.device_ids) > 1:
+                    # intra-node buffer sync
+                    result = broadcast_coalesced(self.modules_buffers_data[0],
+                                                 self.device_ids,
+                                                 self.broadcast_bucket_size)
+                    for tensors, module_buffers_data in zip(result[1:], self.modules_buffers_data[1:]):
+                        for tensor, buffer_data in zip(tensors, module_buffers_data):
+                            buffer_data.set_(tensor)
+
+    def _register_grad_hooks(self):
+        self._grad_accs = []  # need to keep them in scope
+        for device_idx, module in enumerate(self._module_copies):
+            for p in module.parameters():
+                if p.requires_grad:
+                    p_tmp = p.expand_as(p)
+                    grad_acc = p_tmp.grad_fn.next_functions[0][0]
+                    grad_acc.register_hook(self._make_param_hook(p, device_idx))
+                    self._grad_accs.append(grad_acc)
+
+    def _make_param_hook(self, param, device_idx):
+        bucket_idx = self.bucket_map[param]
+
+        def distributed_data_parallel_hook(*unused):
+            if param.grad.requires_grad:
+                raise RuntimeError("DistributedDataParallelC10d only works "
+                                   "with gradients that don't require grad")
+            bucket = self.buckets[bucket_idx][device_idx]
+            bucket.append(param.grad.data)
+
+            # We can flush these and save memory for replicas
+            if device_idx > 0:
+                param.grad = None
+                param.data.set_()
+
+            # Current device's bucket is full
+            if len(bucket) == self.bucket_sizes[bucket_idx]:
+                self.devs_ready[bucket_idx] += 1
+                if self.devs_ready[bucket_idx] < len(self.device_ids):
+                    return
+
+                if bucket_idx not in self.buckets_ready:
+                    self.buckets_ready.add(bucket_idx)
+                    if len(self.buckets_to_reduce) > 0 and self.buckets_to_reduce[-1] == bucket_idx:
```

Comment:
I know you said it's a safety check, but I really think that having too many "hidden assertions" is very bad for code readability. When I read a condition I always assume it's meaningful. If you want to verify invariants of your algorithm (like emptiness of `buckets_to_reduce`) make it clear that it's not critical logic, and raise an AssertionError if it fails (instead of skipping a branch and silently introducing incorrect behavior).

Reply:
addressed